### ISSUE TITLE
Thread affinity in bandwidth benchmark

### ISSUE THREAD

**Issue by carstenbauer** [giordano/julia-on-fugaku#1]
Not sure if you still care about comments but here we go 😄:

For the [bandwidth benchmark](https://github.com/giordano/julia-on-fugaku/tree/main/benchmarks/bandwidthbenchmarkjl) you have used [`export JULIA_EXCLUSIVE=1`](https://github.com/giordano/julia-on-fugaku/blob/91b07995bd2b63761edb57e6376a7b6b77c59728/benchmarks/bandwidthbenchmarkjl/job.sh#L10) and thus compact pinning for the Julia threads. This explains (most of) the character of the [flops plot](https://github.com/giordano/julia-on-fugaku/blob/main/benchmarks/bandwidthbenchmarkjl/flopsscaling.pdf) (and probably also the [bandwidth plot](https://github.com/giordano/julia-on-fugaku/blob/main/benchmarks/bandwidthbenchmarkjl/bwscaling.pdf)).

This also almost certainly underestimates the achievable flops/bandwidth for a given number of threads (see https://github.com/JuliaPerf/BandwidthBenchmark.jl#compact-vs-scattered-pinning). Instead, I'd recommend to choose a scatter pinning that respects the NUMA structure. You can use ThreadPinning.jl for this (which you already have in the `Project.toml`), specifically `pinthreads(:scatter; places=:numa)`.

**Comment by giordano** [comment]
Yeah, I never came back to these benchmarks after the first run, I'm definitely interested in trying again with your suggestions!

**Comment by giordano** [comment]
How do these benchmarks look: https://github.com/giordano/julia-on-fugaku/tree/cd5795aa746ec83286dc5d82aefdde50c56f74a3/benchmarks/bandwidthbenchmarkjl? :slightly_smiling_face:  (These were run on Ookami, much quicker to get a node here.)  I guess a direct comparison would be https://github.com/RRZE-HPC/TheBandwidthBenchmark/wiki/Fujitsu-FX700-S1-M4-C48, except that one doesn't show SDaxpy scaling, which is what we have here instead.

**Comment by carstenbauer** [comment]
Yes, that looks much better. Also, it seems to give O(700GB/s) which is reasonable I guess (ist it really GB or GiB?).

cc @vchuravy ☝️ Not sure why you did only get ~250GB/s in your webinar live demo. OTOH, looking at the RRZE benchmarks there seems to be quite some variation between different kernels.

**Comment by giordano** [comment]
> (ist it really GB or GiB?).

Whatever BandwidthBenchmarks spits out, which according to the output is MB/s, not MiB/s.

> OTOH, looking at the RRZE benchmarks there seems to be quite some variation between different kernels.

Is there a way to choose the kernel for the scaling plots in BandwidthBenchmarks? We only have sdaxpy, which is the only one not reported by RRZE, so making a comparison is hard. Also, they group processors by numa domains, which I guess avoids the ups and downs.

**Comment by carstenbauer** [comment]
> Is there a way to choose the kernel for the scaling plots in BandwidthBenchmarks?

No, it's currently [hard-coded](https://github.com/JuliaPerf/BandwidthBenchmark.jl/blob/main/src/bwscaling.jl) but could easily be extended in this direction. Feel free to open an issue / a PR.
